eq3btsmart: support modes and clean up the code to use climatedevice's features

rytilahti opened the pull request:
**Description:** This patch adds support for modes of eq3 bluetooth smart thermostat.

~~Note: this requires an updated fork of the backend lib (PR https://github.com/bimbar/bluepy_devices/pull/6). The upstream hasn't yet responded.~~

Update: As the upstream did not answer, I decided to clean up the library, fix some issues and upload it into PyPI. This is now ready to be merged.

Update2: The device connections are also not hold anymore, but connections are done when required, so using other means for controlling do also work. Besides that, an issue where the helper util hangs is fixed in the new cleaned up lib.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#1618


**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51

rytilahti committed:
eq3btsmart: support modes and clean up the code to use climatedevice's features

jannau commented on homeassistant/components/climate/eq3btsmart.py:
```diff
@@ -81,24 +102,41 @@ def set_temperature(self, **kwargs):
-        return {
-            ATTR_MODE: self._thermostat.mode,
-            ATTR_MODE_READABLE: self._thermostat.mode_readable,
-        }
```
Reporting the mode as attribute is quite pointless now that we represent it as device state but there are some values the updated eq3btsmart which should be reported as attributes:
eq3.valve_state (homematic.py reports it as 'Valve'), eq3.locked (nest reports that via a binary sensor as 'is_locked'), eq3.window_open and eq3.low_battery.

rytilahti replied:
Done. They don't seem to be visible anywhere in homessistant at this point though.

jannau replied:
they are visible in the log and in web frontend under /dev-state. They are also useable in automation rules like notification on low battery.

rytilahti replied:
Indeed, I forgot about /dev-state completely. They are there now and it seems to work, thanks.

pvizeli replied:
We make a cleanup and set all to lower case on homematic.

rytilahti replied:
Hmm, github is showing this code here even though a newer commit re-adds device_state_attributes?! See https://github.com/home-assistant/home-assistant/pull/4959/commits/55bed24f8333354bdd7d3aa08ee112baa76b1f72 :-)

rytilahti committed:
eq3btsmart: re-add device state attributes

adds reporting for is_locked, valve, window_open and low_battery,
exposing everything the device reports currently.

jannau commented:
I'm using this since the day the pull request was created and I haven't noticed any problems.

@bimbar, the maintainer of bluepy_devices (on github and pypi) is unfortunately MIA so I guess we should use my or @rytilahti github repo as requirement

rytilahti commented:
I have also been using this as long as I remember, since today also with more thermostat, without problems. I'd propose we create a new project only for eq3bt, fork away relevant pieces of code (we don't need all the library code for our use, so we can just take what's needed and not depend on bluepy_devices at all. What do you think?

rytilahti committed:
eq3btsmart: bump version req

rytilahti commented:
@jannau, just FYI: I started to refactor the lib to move eq3btsmart-specific lib only. I also found out (probably) why the backend lib was left hanging, making restart delays of homeassistant unberable and I haven't encountered yet any problems. The code is not pushed anywhere yet, I can do that when it's time.

rytilahti committed:
eq3btsmart: fix a typo in mode name, report unknown when not initialized

rytilahti committed:
eq3btsmart: depend on newly forked python-eq3bt lib, pythonify states

balloob commented on homeassistant/components/climate/eq3btsmart.py:
```diff
@@ -81,24 +108,53 @@ def set_temperature(self, **kwargs):
+    @property
+    def operation_list(self):
+        """List of available operation modes."""
+        return [x for x in self.modes.values()]
```
For next time, this is equivalent: `list(self.modes.values())`

balloob commented:
Looks good! :dolphin:

bimbar commented:
Great, can't wait for the new features.
